### PR TITLE
[Merged by Bors] - chore: move `mkRichHCongr` from `congr!` implementation into CongrTheorems module

### PR DESCRIPTION
**Pull request by kmill**
1. Defines a trace class `Meta.CongrTheorems` for congruence theorem generators in general
2. Renames `Congr!.mkHCongrThm` to `Lean.Meta.mkRichHCongr` for this congruence theorem generator
3. Adds a `forceHEq` parameter to the function; `false` is the default and makes it act like `Lean.Meta.mkHCongr`, but `congr!` expects the lemma to prove a `HEq` and not an `Eq`.
4. Makes `mkRichHCongr` package up some more data and return a `CongrTheorem`

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by kmill** [comment]
@semorrison Here's another meta moving PR. This will make it a bit easier to try to refactor the two congruence generators that are now in the CongrTheorems module.

File: Mathlib/Lean/Meta/CongrTheorems.lean [hunk 1]
```diff
@@ -69,4 +71,192 @@ where
     -- We have no more tricks.
     failure
 
+/--
+Create a congruence lemma to prove that `Eq/HEq (f a₁ ... aₙ) (f' a₁' ... aₙ')`.
```
Comment (semorrison) [inline Mathlib/Lean/Meta/CongrTheorems.lean:75]: For a long doc-string, I like to start with something like "`mkRichHCongr fType info fixedFun fixedParams forceHEq` creates ...", so the name of the function and the arguments are immediately in the reader's input context.

File: test/congr.lean [hunk 2]
```diff
@@ -6,6 +6,8 @@ import Mathlib.Data.List.Defs
 
 set_option autoImplicit true
 
+--set_option trace.Meta.CongrTheorems true
+
```
Comment (semorrison) [inline test/congr.lean:10]: Either remove entirely, or add a line `-- This option is useful for debugging:`

File: scripts/nolints.json [hunk 3]
```diff
@@ -2160,6 +2157,8 @@
  ["docBlame", "Lean.Meta.Simp.mkCast"],
  ["docBlame", "Lean.Meta.Simp.mkEqSymm"],
  ["docBlame", "Lean.Meta.findCore.check"],
+ ["docBlame", "Lean.Meta.mkRichHCongr.doubleTelescope.loop"],
+ ["docBlame", "Lean.Meta.mkRichHCongr.withNewEqs.loop"],
```
Comment (semorrison) [inline scripts/nolints.json:2161]: How about we just write the doc-strings, even if they are just `implementation of ...`?

**Comment by semorrison** [comment]
bors d+

File: scripts/nolints.json (hunk 3 above)
Comment (kmill) [reply scripts/nolints.json:2161]: Hopefully one day Lean will tag local declarations that come from things like `let rec` to help the linter...

**Pull request closed without merging**